Enable lowering to inductor

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97431
* #97122
* #97133
* #97418
* #97132
* #96475
* __->__ #96927

**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

leslie-fang-intel committed:
Enable lowering to inductor

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@jansel @jerryzh168 @supriyar @vkuzo Could you help to take a look of this PR?

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

EikanWang commented on torch/_inductor/compile_fx.py:
```diff
@@ -457,6 +453,10 @@ def compile_fx(
                 recursive_compile_fx,
             )
 
+        with overrides.patch_functions():
```
Could you please add some comments here to explain why you need to put the code after `handle_dynamo_export_graph`?

leslie-fang-intel replied:
Thanks for the suggestions. The comments are added to explain why we move it after `handle_dynamo_export_graph`.

jansel replied:
Are you sure this is actually run?  Doesn't `handle_dynamo_export_graph` fail the `if isinstance(model_, torch.fx.GraphModule):` check on L466?

leslie-fang-intel replied:
Hi @jansel, Thanks for the comment. From my experiment based on the UT in this PR, I do see it will enter into these passes twice if we put them afront of `handle_dynamo_export_graph`.
1. The first hit is the first call of `compile_fx`.
2. The second hit is `compiled_fn = compile_gm(gm, codegen.process_inputs(*inputs))` inside `handle_dynamo_export_graph` function.

Can you explain more why `handle_dynamo_export_graph` will fail the `if isinstance(model_, torch.fx.GraphModule):`. Based on my understand, looks like `handle_dynamo_export_graph`  just change the `gm.graph._codegen` from `_PyTreeCodeGen` to `CodeGen`. And `gm` is still the `torch.fx.GraphModule`.

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
> Are you sure this is actually run? Doesn't `handle_dynamo_export_graph` fail the `if isinstance(model_, torch.fx.GraphModule):` check on L466?

Hi @jansel, regarding to this comment. I am still not quite understanding of your concerns. Could you please explain more or suggest about how to resolve it?

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel reviewed (approved):
My mistake, I was confusing `handle_dynamo_export_graph` for another one.

leslie-fang-intel commented:
Hi @jerryzh168, could you please kindly take a look of this PR? It relates with the usage API we proposed for the Quantization 2.0 Inductor path.

jerryzh168 commented:
> 2\. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

sure, lowering API LGTM

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel committed:
Update on "Enable lowering to inductor"


**Summary**
Enable the lowering path from a quantized 2.0 fx graph into Inductor. The basic usage will be 
```
export_module, guards = torchdynamo.export(m, *args)
prepare_module = prepare_pt2e(export_module, *args)
convert_module = convert_pt2e(prepare_module)
ooptimized_module = compile_fx(convert_module, example_inputs)
```
Most of the issues we met previously has already been fixed in PyTorch Master. So in this PR, we mainly do 2 things:
1. Add the basic usage into a UT.
2. Move `handle_dynamo_export_graph` before the fusion passes, otherwise the dynamo_export_graph will hit the fusion passes twice which is un-expected.

**Test Plan**
```
clear && python -m pytest test_quantization.py -k test_inductor_backend_config_conv
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
